### PR TITLE
added schema for all parameters

### PR DESCRIPTION
**Pull request by fehguy**
This is NOT a complete change, but conceptually it loosens the requirement that all parameters except `body` parameters be of a simple type.  Instead, a schema can be provided for all parameter types.

For parameter types of type `body`, the schema will be serialized by the `consumes` attribute.  For all other parameter types, a serialization strategy needs to be defined (and is still TBD for definition).

This serialization strategy would be responsible for turning a complex object into JSON, for example.

### CODE REVIEW

**Comment by fehguy** [comment]
This addresses #412, #437, #69, #401, #179 (by inclusion of the `schema`), #301 

For issue #222, we can also add a format for `file` to indicate the encoding type.

For issue #300, since schemas can be reused, this should address that as well.

This solves #595 if we can define the serialization strategy

File: versions/3.0.md [hunk 1]
```diff
@@ -599,35 +601,7 @@ Field Name | Type | Description
...
-<a name="parameterMaxLength"></a>maxLength | `integer` | See http://json-schema.org/latest/json-schema-validation.html#anchor26.
-<a name="parameterMinLength"></a>minLength | `integer` | See http://json-schema.org/latest/json-schema-validation.html#anchor29.
-<a name="parameterPattern"></a>pattern | `string` | See http://json-schema.org/latest/json-schema-validation.html#anchor33.
-<a name="parameterMaxItems"></a>maxItems | `integer` | See http://json-schema.org/latest/json-schema-validation.html#anchor42.
-<a name="parameterMinItems"></a>minItems | `integer` | See http://json-schema.org/latest/json-schema-validation.html#anchor45.
-<a name="parameterUniqueItems"></a>uniqueItems | `boolean` | See http://json-schema.org/latest/json-schema-validation.html#anchor49.
-<a name="parameterEnum"></a>enum | [*] | See http://json-schema.org/latest/json-schema-validation.html#anchor76.
-<a name="parameterMultipleOf"></a>multipleOf | `number` | See http://json-schema.org/latest/json-schema-validation.html#anchor14.
+<a name="parameterSchema"></a>schema | [Schema Object](#schemaObject) | The schema defining the type used for the parameter.
```
Comment (ePaul) [inline versions/3.0.md]: Allow a reference object (i.e. a reference to a schema object) here too?

File: versions/3.0.md [hunk 2]
```diff
@@ -713,11 +689,12 @@ name: token
...
-items:
-  type: integer
-  format: int64
-collectionFormat: csv
+schema:
+  type: array
+  items:
+    type: integer
+    format: int64
+  collectionFormat: csv
```
Comment (fehguy) [inline versions/3.0.md]: with this change, `collectionFormat` may not make sense.  It should be added to the serialization strategy

**Comment by fehguy** [comment]
need to reconsider this when looking at 6570 support

**Comment by fehguy** [comment]
@OAI/tdc please post other comments here--i will not merge until a serialization strategy is provided and there is more time for feedback

File: versions/3.0.md (hunk 1 above)
Comment (webron) [inline versions/3.0.md]: @ePaul - The `Schema Object` allows a `$ref` as its value, so there's no need to be explicit about it. It's the same as how `body` parameters behave today.

**Comment by webron** [comment]
This does _not_ solve #69 nor #401.

To me the serialization strategy can be controlled by a `consumes` property at each parameter, which will override any upper-level `consumes` value (either at the operation or OpenAPI Object level). This will also help resolve #222. This gets tricky since primitives cannot really be serialized as `application/json` for example, so we may explicitly mention that primitives are passed as-is.

To me, the main issue is arrays and the existing `collectionFormat`. No mime-type will actually cover the existing cases in it. Brain is a bit fried right now, the only solution I can come up with right now is fairly ugly - will try to think about it more later.

I do _not_ advocate using complex models for query/path/header parameters (body and form are okay) - but in favor of simplicity I :+1: the concept. Just need to work through the open questions.

**Comment by fehguy** [comment]
* Move type definitions out of serializable parameters
* Add a `schema` to all
* come up with a serialziation strategy for all non-body parameters

**Comment by Vanuan** [comment]
Is there any rough implementation?

I'm looking for a simple query parameter of type object serialized into a JSON string.

**Comment by ePaul** [comment]
@Vanuan I'm not aware of any. We are currently working on the next version of the specification (3.0), and only if that it out, the tools can reliably be updated to support it. (I guess some tools will start implementation earlier, though.)

**Comment by Vanuan** [comment]
Is there some criteria or time frame for 3.0 to be released?
I mean, are you working towards supporting describing specific APIs, features or do you plan to release new spec each year?
For example you could say that in 3.0 you'd want to support describing such APIs:

- Facebook https://developers.facebook.com/tools/explorer/
- Twitter
- Sailsjs blueprint API http://sailsjs.org/documentation/reference/blueprint-api

Or you could say that you target supporting specific features, like an ability to specify fields like this:
?fields=id,name
or specifying objects in queries like this:
?filter={"name":{"includes":"john"}}

Is GitHub issues is the only place or is there some highlevel roadmap with milestone?
In Readme it's mentioned that OpenAPI doesn't plan to support every REST API use case. Does it mean that applications would need to adapt to OpenAPI features to fully employ its benefits?

**Comment by webron** [comment]
@Vanuan we're aiming for a summer release for 3.0. The changes are done via specific feature requests and not aiming to support specific APIs.

Github issues is the place.

It means that not all APIs could be described with OpenAPI. Whether applications change their APIs or don't use OpenAPI is their call.

**Pull request merged**